### PR TITLE
Driver: Implement Expand Volume

### PR DESCRIPTION
**Pull request by pierre-emmanuelJ**
# Description
Add the ability to resize a Block Storage volume and expand the file-system.

## Checklist
(For exoscale contributors)

* [x] Changelog updated (under *Unreleased* block)
* [x] Integration tests OK

### CODE REVIEW

File: driver/helpers.go [hunk 1]
```diff
@@ -94,3 +95,53 @@ func createMountPoint(path string, file bool) error {
 func convertBytesToGibiBytes(nBytes int64) int64 {
 	return nBytes / (1024 * 1024 * 1024)
 }
+
+func getNewVolumeSize(capacityRange *csi.CapacityRange) (int64, error) {
```
Comment (sauterp) [inline driver/helpers.go:99]: Could you write some tests for this function? (Not integ tests just some "unit" tests to verify the logic)

File: driver/helpers.go [hunk 2]
```diff
@@ -94,3 +95,53 @@ func createMountPoint(path string, file bool) error {
...
+
+	if requiredSet && requiredBytes > MaximumVolumeSizeBytes {
+		return 0, errors.New("required size is greater than the maximum size")
+	}
+
+	if !requiredSet && limitSet && limitBytes > MaximumVolumeSizeBytes {
+		return 0, errors.New("limit size is greater than the maximum size")
+	}
+
+	if requiredSet && limitSet && requiredBytes == limitBytes {
```
Comment (sauterp) [inline driver/helpers.go:134]: I think this case is unnecessary given the next one.

File: driver/controller.go [hunk 3]
```diff
@@ -605,8 +607,54 @@ func (d *controllerService) ListSnapshots(ctx context.Context, req *csi.ListSnap
...
+	if volumeCapability != nil {
+		err := validateVolumeCapability(volumeCapability)
+		if err != nil {
+			return nil, status.Errorf(codes.InvalidArgument, "volumeCapabilities not supported: %s", err)
+		}
+
+		if _, ok := volumeCapability.GetAccessType().(*csi.VolumeCapability_Block); ok {
+			nodeExpansionRequired = false
+		}
+	}
```
Comment (sauterp) [inline driver/controller.go:652]: wouldn't it be better to do these checks before we do the resize call?
As it is now the block storage volume could get resized and then would you could get the error "volumeCapabilities not supported..." after.

**Review by sauterp — COMMENTED** [review]
don't forget to add an integration test please

**Comment by zifeo** [comment]
hello @pierre-emmanuelJ, we would be interested in testing that feature. Is the PR safe enough to be tested/used?

**Comment by pierre-emmanuelJ** [comment]
> hello @pierre-emmanuelJ, we would be interested in testing that feature. Is the PR safe enough to be tested/used?

Hello @zifeo 

Thanks for your interest in the Block Storage volume resize feature! 

Unfortunately, due to limitations in the current Exoscale Block Storage design, this PR currently only supports offline resizing. This means that to resize a volume, it would need to be detached from the Kubernetes node first.

While we understand this isn't ideal for a Kubernetes environment... 

Online resizing isn't currently supported by Exoscale. This PR implements the groundwork for future online resizing when Exoscale offers that functionality.

**Important Note**: This PR currently support offline volume resizing only. To resize a volume, it must be detached from the Kubernetes node, causing downtime for pods using that volume.

**Comment by zifeo** [comment]
@pierre-emmanuelJ Thanks for the notice, is there a place where we can track the progress or a roadmap for online resizing?

**Comment by sauterp** [comment]
@zifeo I created an issue to track the issue of online resizing, any updates will be posted there. https://github.com/exoscale/exoscale-csi-driver/issues/23

**Review by sauterp — APPROVED** [review]

**Review by sauterp — CHANGES_REQUESTED** [review]
integ-test isn't passing yet :(

**Review by sauterp — APPROVED** [review]

**Review by arnaudgeiser — APPROVED** [review]

**Pull request merged**